### PR TITLE
Use react hooks in TagsInput

### PR DESCRIPTION
**Pull request by stellarhoof**
Use an observable local store and add a storybook for TagsInput. Also fixed an issue with the stories not being able to use hooks.

### CODE REVIEW

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  This PR has failing tests. Please alleviate the errors and commit them

<details>
  <summary><a href="https://github.com/smartprocure/contexture-react/blob/feature/tags-input/s" target="_blank">s</a></summary>
  <code>  ● Storyshots › Components (Unthemed)|Layout › TagsInput

    expect(received).toMatchSnapshot()

    Snapshot name: `Storyshots Components (Unthemed)|Layout TagsInput 1`

    - Snapshot
    + Received

    @@ -3,36 +3,186 @@
          Object {
            "fontSize": 20,
          }
        }
      >
    +   <h2>
    +     Doesn't rerender on different props
    +   </h2>
        <div
          style={
            Object {
              "display": "flex",
            }
          }
        >
          <button
            onClick={[Function]}
          >
    -       Rerender
    +       Cycle props
          </button>
          <button
            onClick={[Function]}
          >
    -       Toggle periodical rerender
    +       Toggle periodical props cycling
          </button>
        </div>
        <div
          style={
            Object {
              "margin": 10,
            }
          }
        >
    -     Rendered with props: {"initialTags":["tag1","tag2"]}
    +     Props: {"initialTags":["tag1","tag2"]}
    +   </div>
    +   <div
    +     className="tags-input"
    +     style={
    +       Object {
    +         "height": "100%",
    +       }
    +     }
    +   >
    +     <div
    +       style={
    +         Object {
    +           "alignItems": "center",
    +           "cursor": "text",
    +           "display": "flex",
    +           "flexWrap": "wrap",
    +           "height": "100%",
    +           "padding": 3,
    +         }
    +       }
    +     >
    +       <span
    +         className="tags-input-tag"
    +         onClick={[Function]}
    +         style={
    +           Object {
    +             "backgroundColor": "#0076de",
    +             "borderRadius": "3px",
    +             "color": "white",
    +             "cursor": "pointer",
    +             "margin": 3,
    +             "padding": 5,
    +           }
    +         }
    +       >
    +         <div
    +           style={
    +             Object {
    +               "alignItems": "center",
    +               "display": "flex",
    +             }
    +           }
    +         >
    +           <span
    +             style={
    +               Object {
    +                 "paddingBottom": "0.15em",
    +                 "paddingLeft": "0.45em",
    +                 "paddingRight": "0em",
    +               }
    +             }
    +           >
    +             tag1
    +           </span>
    +           <span
    +             className="tags-input-tag-remove"
    +             onClick={[Function]}
    +           >
    +             x
    +           </span>
    +         </div>
    +       </span>
    +       <span
    +         className="tags-input-tag"
    +         onClick={[Function]}
    +         style={
    +           Object {
    +             "backgroundColor": "#0076de",
    +             "borderRadius": "3px",
    +             "color": "white",
    +             "cursor": "pointer",
    +             "margin": 3,
    +             "padding": 5,
    +           }
    +         }
    +       >
    +         <div
    +           style={
    +             Object {
    +               "alignItems": "center",
    +               "display": "flex",
    +             }
    +           }
    +         >
    +           <span
    +             style={
    +               Object {
    +                 "paddingBottom": "0.15em",
    +                 "paddingLeft": "0.45em",
    +                 "paddingRight": "0em",
    +               }
    +             }
    +           >
    +             tag2
    +           </span>
    +           <span
    +             className="tags-input-tag-remove"
    +             onClick={[Function]}
    +           >
    +             x
    +           </span>
    +         </div>
    +       </span>
    +       <input
    +         onBlur={[Function]}
    +         onKeyDown={[Function]}
    +         placeholder="Search..."
    +         style={
    +           Object {
    +             "border": "none",
    +             "flex": 1,
    +             "margin": 3,
    +             "outline": "none",
    +           }
    +         }
    +       />
    +     </div>
    +   </div>
    +   <h2>
    +     Rerenders on different props
    +   </h2>
    +   <div
    +     style={
    +       Object {
    +         "display": "flex",
    +       }
    +     }
    +   >
    +     <button
    +       onClick={[Function]}
    +     >
    +       Cycle props
    +     </button>
    +     <button
    +       onClick={[Function]}
    +     >
    +       Toggle periodical props cycling
    +     </button>
    +   </div>
    +   <div
    +     style={
    +       Object {
    +         "margin": 10,
    +       }
    +     }
    +   >
    +     Props: {"deriveStateFromProps":true,"initialTags":["tag1","tag2"]}
        </div>
        <div
          className="tags-input"
          style={
            Object {

      at match (node_modules/@storybook/addon-storyshots/dist/test-bodies.js:27:20)
      at node_modules/@storybook/addon-storyshots/dist/test-bodies.js:39:10
      at Object.<anonymous> (node_modules/@storybook/addon-storyshots/dist/api/snapshotsTestsTemplate.js:35:33)
</code>
</details>
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>The README has not been updated. Please update the README.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Could not find any browser results.</td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against f98c1a611e1fa6b3fc3108b183ae84248e7ecb3b
</p>

**Comment by stellarhoof** [comment]
Having issues with using contexture-react with hooks in bid-search.

File: stories/layout.stories.js [hunk 1]
```diff
@@ -42,6 +46,42 @@ let HighlightDemo = withStateLens({ filter: '' })(
   ))
 )
 
+let MockTagsInput = ({ initialTags, deriveStateFromProps = false }) => {
+  let [tags, setTags] = useState(initialTags)
+  if (deriveStateFromProps && initialTags !== tags) setTags(initialTags)
```
Comment (daedalus28) [inline stories/layout.stories.js]: this is confusing :)

File: stories/layout.stories.js [hunk 2]
```diff
@@ -42,6 +46,42 @@ let HighlightDemo = withStateLens({ filter: '' })(
...
+      tags={tags}
+      addTag={t => setTags(_.concat(tags, t))}
+      removeTag={t => setTags(_.pull(t, tags))}
+      tagStyle={{ color: 'white', backgroundColor: '#0076de', padding: 5 }}
+      PopoverContents={({ tag }) => <div>{tag}</div>}
+    />
+  )
+}
+
+let DynamicProps = ({ Component, propList }) => {
```
Comment (daedalus28) [inline stories/layout.stories.js]: This is a little hard to follow - why do we need a demo with setInterval?

File: src/utils/react.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { useState, useEffect, useRef } from 'react'
+
+export let useStateLens = value => {
+  let [state, setState] = useState(value)
+  return { get: () => state, set: setState }
+}
+
+// https://overreacted.io/making-setinterval-declarative-with-react-hooks/
+export let useInterval = (callback, delay = 1000) => {
```
Comment (daedalus28) [inline src/utils/react.js]: This is weird and only used for one story, so probably shouldn't be part of the lib's utils

File: src/utils/react.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { useState, useEffect, useRef } from 'react'
+
+export let useStateLens = value => {
```
Comment (daedalus28) [inline src/utils/react.js]: if we're gonna do this, let's make sure it's exported properly from the index as well and just make this a thing

File: src/utils/react.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  let [state, setState] = useState(value)
+  return { get: () => state, set: setState }
```
Comment (daedalus28) [inline src/utils/react.js]: if you rename `setState` to `set`, you can use object shorthand here

File: src/utils/react.js (hunk 4 above)
Comment (daedalus28) [reply src/utils/react.js]: But I still think we should just make popover/modal/whatever not take an actual lens as part of the API

File: stories/layout.stories.js (hunk 2 above)
Comment (stellarhoof) [reply stories/layout.stories.js]: The idea was to test if the input lost focus on rerenders. It kept it in the story but it didn't on our main product, so it's ok to remove.

File: package.json [hunk 6]
```diff
@@ -102,7 +102,7 @@
     "storybook-addon-jsx": "^7.1.0"
   },
   "dependencies": {
-    "futil-js": "^1.55.0",
+    "futil-js": "^1.56.0",
```
Comment (stellarhoof) [inline package.json]: Whoops. Might as well leave it here though

**Comment by stellarhoof** [comment]
@daedalus28 Do not merge yet. Latest mobx-react breaks the app if we npm link contexture-react to it.

**Comment by stellarhoof** [comment]
Blocked on https://github.com/smartprocure/contexture-react/pull/259

**Comment by daedalus28** [comment]
Closing in favor of #275
